dockerplugin.py

```diff
@@ -611,14 +715,43 @@ def read_callback(self):
                 if not read_at:
                     # No stats available yet; skipping container.
                     continue
-
+                if self.cpu_shares_bool or self.cpu_quota_bool:
+                    # Get cgroup info container by inspecting the container
+                    container_inspect = self.client \
+                                                .inspect_container(container)
+                    cpu_shares = container_inspect['HostConfig']['CpuShares']
+                    # Calculate the sum of weights of all containers
+                    sum_of_shares += (cpu_shares or DEFAULT_SHARES)
+                    # Store state of containers for second pass (greedy stats)
+                    containers_state[container['Name']] = {
+                                'container': container,
+                                'cstats': cstats,
+                                'stats': stats,
+                                'read_at': read_at,
+                                'container_inspect': container_inspect
+                    }
                 # Process stats through each reader.
                 for method in self.METHODS:
                     method(container, cstats.dimensions, stats, read_at)
+
             except Exception, e:
                 log.exception(('Unable to retrieve stats for container '
                                '{container}: {msg}')
                               .format(container=_c(container), msg=e))
+                              
+        if self.cpu_quota_bool or self.cpu_shares_bool:
```

Comment:
So, my understanding is that you have to do this is two passes because you need to calculate `sum_of_shares` before you can compute something in `read_greedy_container_stats`. It seems like you only need `sum_of_shares` if you've been asked to calculate the `cpu.shares` metrics.

Given this, I think you should split the reading of greedy container stats into two methods, one for CPU quota and one for CPU shares, only calling each method if you need to. I would also calculate the sum of shares and call those methods later on, to keep the main loop simple:

In the main loop:

```python
# Process stats through each reader
for method in self.METHODS:
    method(container, cstats.dimensions, stats, read_at)

if self.cpu_shares_bool or self.cpu_quota_bool:
    # Gather detailed container info for later
    containers_state[container] = self.client.inspect_container(container)
```

Then, below:

```python
sum_of_shares = reduce(
    lambda a, b: a + b, map(
        lambda i: i['HostConfig']['CpuShares'] or 1024),
        containers_state.values), 0)
for container, state in containers_state.items():
    cstats = self.stats[container['Id']]
    cpu_percent = get_cpu_percent(cstats.stats)
    if self.cpu_quota_bool:
        read_cpu_quota_stats(container, state, cstats, cpu_percent)
    if self.cpu_shares_bool:
        read_cpu_shares_stats(container, state, cstats, cpu_percent, sum_of_shares)
```

Reply:
I'll make this change. This looks so much more cleaner.